File: Session14/1268. Search Suggestions System.java
class Solution {
    class TrieNode
    {
        TrieNode children[];
        boolean isEnd;
        public TrieNode(){
            children=new TrieNode[26];
            isEnd=false;
            for(int i=0;i<26;i++)
            {
                children[i]=null;
            }
        }
    }
    
    List<String> res;
    
    public void dfsWithPrefix(TrieNode curr,String prefix)
    {
        if(res.size()==3)
        {
            return;
        }
        
        if(curr.isEnd==true)
        {
            res.add(prefix);
        }
        
        for(char c='a';c<='z';c++)
        {
            if(curr.children[c-'a']!=null)
            {
                dfsWithPrefix(curr.children[c-'a'],prefix+c);
            }
        }
    }
    
    
    public List<String> getSearchResult(String prefix)
    {
       res=new ArrayList<String>();
        TrieNode curr=root;
        for(char p:prefix.toCharArray())
        {
            if(curr.children[p-'a']!=null)
            {
                curr=curr.children[p-'a'];
            }
            else
            {
                return new ArrayList<String>();
            }
        }
        dfsWithPrefix(curr,prefix);
        return res;
    }
    
    
    TrieNode root;
    public List<List<String>> suggestedProducts(String[] products, String searchWord) {
        root=new TrieNode();
        List<List<String>> ans=new ArrayList<>();
        
        for(String product:products)
        {
            insert(product);
        }
        
        String prefix=new String();
        for(char w:searchWord.toCharArray())
        {
            prefix=prefix+w;
            ans.add(getSearchResult(prefix));
        }
      return ans;
    }
    
    public void insert(String word) {
    //System.out.println("ins");
    TrieNode curr=root;
    for(char w:word.toCharArray())
    {
        if(curr.children[w-'a']!=null)
        {
            curr=curr.children[w-'a'];
        }
        else
        {
            curr.children[w-'a']=new TrieNode();
            curr=curr.children[w-'a'];
        }
    }
    curr.isEnd=true;
    }
}

File: Session14/1268. Search Suggestions System.py
class TrieNode(object):
    def __init__(self, val):
        self.val = val
        self.children = collections.OrderedDict()
        self.isEnd = False

class Trie(object):
    def __init__(self):
        self.root = TrieNode(None)
    
    def insert(self, word):
        node = self.root
        for char in word:
            if char not in node.children.keys():
                node.children[char] = TrieNode(char)
            node = node.children[char]
        
        node.isEnd = True
    
    def getSuggestions(self, searchWord):
        suggestions = []
        node = self.root
        for char in searchWord:
            if char not in node.children.keys():
                return suggestions
            node = node.children[char]
        
        suggestions = self.dfs(node, searchWord[:-1], "", [])
        return suggestions
    
    def dfs(self, node, prefix, path, res):
        if len(res) >= 3:
            return res
        
        path += node.val
        
        if node.isEnd:
            res.append(prefix + path[:])
        
        for key, child in node.children.items():
            res = self.dfs(child, prefix, path[:], res)
        
        return res
                
        
class Solution(object):
    def suggestedProducts(self, products, searchWord):
        """
        :type products: List[str]
        :type searchWord: str
        :rtype: List[List[str]]
        """
        
        trie = Trie()
        products.sort()
        for product in products:
            trie.insert(product)
        res = []
        curr_searchWord = ""
        for char in searchWord:
            curr_searchWord += char
            suggestions = trie.getSuggestions(curr_searchWord)
            res.append(suggestions)
        
        return res